terraform/iam.tf

```diff
@@ -0,0 +1,73 @@
+#
+# Copyright 2021, Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+resource "google_project_iam_custom_role" "terraform_service_account_role" {
+  project = var.project
+  role_id = "ENPATerraformRunner"
+  title   = "Terraform Runner for ENPA Infra provisioning"
+  permissions = [
+    "cloudkms.cryptoKeyVersions.create",
+    "cloudkms.cryptoKeyVersions.destroy",
+    "cloudkms.cryptoKeyVersions.get",
+    "cloudkms.cryptoKeyVersions.viewPublicKey",
+    "cloudkms.cryptoKeys.create",
+    "cloudkms.cryptoKeys.get",
+    "cloudkms.cryptoKeys.update",
+    "cloudkms.keyRings.create",
+    "cloudkms.keyRings.get",
+    "cloudscheduler.jobs.create",
+    "cloudscheduler.jobs.delete",
+    "cloudscheduler.jobs.get",
+    "firebase.projects.get",
+    "firebaserules.releases.list",
+    "firebaserules.releases.update",
+    "firebaserules.rulesets.create",
+    "firebaserules.rulesets.delete",
+    "firebaserules.rulesets.get",
+    "firebaserules.rulesets.test",
+    "iam.roles.create",
+    "iam.roles.delete",
+    "iam.roles.get",
+    "iam.roles.list",
+    "iam.roles.update",
+    "iam.serviceAccounts.actAs",
+    "iam.serviceAccounts.create",
+    "iam.serviceAccounts.delete",
+    "iam.serviceAccounts.get",
+    "resourcemanager.projects.get",
+    "resourcemanager.projects.getIamPolicy",
+    "resourcemanager.projects.setIamPolicy",
+    "serviceusage.services.disable",
+    "serviceusage.services.enable",
+    "serviceusage.services.get",
+    "serviceusage.services.list",
+    "storage.buckets.create",
+    "storage.buckets.delete",
+    "storage.buckets.get",
+  ]
+}
+
+resource "google_project_iam_member" "terraform_service_account_permissions" {
+  project = var.project
+  role    = "projects/${var.project}/roles/${google_project_iam_custom_role.terraform_service_account_role.role_id}"
+  member  = "serviceAccount:${var.terraform_svc_account_email}"
+}
+
+resource "google_project_iam_binding" "owners" {
+  project = var.project
+  role    = "roles/owner"
+  members = var.project_owners_list
+}
```

Comment:
I'm inferring that the intention of this section is to remove the terraform SA from the project's owners. Is that accurate?

In practice, for our environment, this should work fine. In principle I am nervous about this approach and would prefer to "manually" (like via `gcloud`) remove this account from each project's owners after applying the rest of the config.

I'm nervous because using the authoritative form (`_binding`) on the owners role is an extremely risky footgun to leave laying around. It would be very easy to make a mistake with this resource that could lock us out of all of our projects.

Is it likely? No. Is it possible? Yes. I'd prefer not to take the risk.

Reply:
The org/folder owners stay in control which can add owners back to the project. I agree with the risk, yet would say it is mitigated by having folder level iam permissions to address this. Thus the risk of that vs "forgetting" to remove owner rights from the terraform svc account is weighed in favor to do it here. I wish I could do an operation where I only take away the owner rights from the svc account but couldn't find one other than this approach. Thoughts?

Reply:
Ah, good call on the inherited owners. You're right that mitigates the risk substantially.

This is a one time operation, right? To undo the previous decision to make the terraform SA an owner? If so it still might make sense to do the cleanup as a similar one time operation. You could either do it with gcloud or by running this config once and then removing this block.

Regardless, I feel more comfortable with it now and am totally fine to proceed as you see fit.

Reply:
It applies to each new PHA we setup. So removing would unfortunately not work here.